iml-gui/crate/src/components/action_dropdown.rs

```diff
@@ -0,0 +1,216 @@
+use crate::{
+    components::{dropdown, font_awesome, tooltip, Placement},
+    generated::css_classes::C,
+    WatchState,
+};
+use futures::{
+    channel::oneshot,
+    future::{self, Either},
+    Future, FutureExt,
+};
+use gloo_timers::future::TimeoutFuture;
+use iml_wire_types::{ApiList, AvailableAction, CompositeId, Label, ToCompositeId};
+use seed::{prelude::*, *};
+
+pub type AvailableActions = ApiList<AvailableAction>;
+
+pub trait ActionRecord: ToCompositeId + serde::Serialize + Label + Clone {}
+impl<T: ToCompositeId + serde::Serialize + Label + Clone> ActionRecord for T {}
+
+pub fn composite_ids_to_query_string(xs: &[CompositeId]) -> String {
+    let mut xs: Vec<String> = xs
+        .iter()
+        .map(|x| format!("composite_ids={}", x))
+        .collect::<Vec<String>>();
+
+    xs.sort();
+    xs.join("&")
+}
+
+#[derive(Clone)]
+pub struct IdMsg(pub u32, pub Msg);
+
+#[derive(PartialEq, Eq)]
+pub enum State {
+    Inactive,
+    Activating,
+    Active,
+}
+
+impl Default for State {
+    fn default() -> Self {
+        State::Inactive
+    }
+}
+
+#[derive(Default)]
+pub struct Model {
+    pub state: State,
+    pub composite_ids: Vec<CompositeId>,
+    pub request_controller: Option<fetch::RequestController>,
+    pub actions: Vec<AvailableAction>,
+    pub watching: WatchState,
+    pub cancel: Option<oneshot::Sender<()>>,
+}
+
+impl Model {
+    pub fn new(composite_ids: Vec<CompositeId>) -> Self {
+        Model {
+            state: State::default(),
+            composite_ids,
+            request_controller: None,
+            actions: vec![],
+            watching: WatchState::default(),
+            cancel: None,
+        }
+    }
+}
+
+impl Drop for Model {
+    fn drop(&mut self) {
+        self.cancel = None;
+
+        if let Some(c) = self.request_controller.take() {
+            c.abort();
+        }
+    }
+}
+
+#[derive(Clone)]
+pub enum Msg {
+    StartFetch,
+    SendFetch,
+    WatchChange,
+    Fetched(fetch::ResponseDataResult<AvailableActions>),
+    Noop,
+}
+
+pub fn update(msg: IdMsg, model: &mut Model, orders: &mut impl Orders<IdMsg>) {
+    let IdMsg(id, msg) = msg;
+
+    match msg {
+        Msg::StartFetch => {
+            if model.state == State::Inactive {
+                model.state = State::Activating;
+                orders.send_msg(IdMsg(id, Msg::SendFetch));
+            }
+        }
+        Msg::SendFetch => {
+            model.cancel = None;
+
+            // FIXME: Need to figure out differing urls for dev / prod
+            let request = fetch::Request::new(format!(
+                "https://localhost:7444/api/action/?limit=0&{}",
```

Comment:
I've seen in Seed's example, that URL can be "relative" - `"/api/action/?limit=0&{}"`.

Reply:
Yeah, the problem is the local port isn't matching the remote port right now (due to us using CORS in dev), so using a relative URL won't map to 7444 (8000 locally).